### ISSUE TITLE
Checkpointing is excessively slow, and gets slower with table size, when rows are large

### ISSUE THREAD

**Issue by tmaxwell-anthropic** [duckdb/duckdb#7709]
### What happens?

`CHECKPOINT` sometimes takes time proportional to the size of the table, especially when individual rows are more than a few kilobytes. For example, if I repeatedly insert 1,000 rows of 10,000 bytes each and then `CHECKPOINT`:
1. The first `CHECKPOINT` takes 3 seconds
2. The second `CHECKPOINT` takes 6 seconds 
3. The third `CHECKPOINT` takes 9 seconds
4. And so on

### To Reproduce

```python
import random
import duckdb
import pyarrow
import time
from pathlib import Path


def main(*, n_txns, n_rows_per_txn, n_bytes_per_row):
    db_file = Path("./ducktest.duckdb")
    if db_file.exists():
        db_file.unlink()
    db = duckdb.connect(str(db_file))
    db.execute("CREATE TABLE test (test VARCHAR)")
    for i in range(n_txns):
        data = pyarrow.table(
            {
                "test": pyarrow.array(
                    [random.randbytes(n_bytes_per_row) for i in range(n_rows_per_txn)]
                )
            }
        )
        t0 = time.monotonic()
        db.execute(
            """
            INSERT INTO test
            SELECT test
            FROM data
        """
        )
        t1 = time.monotonic()
        db.execute("CHECKPOINT")
        t2 = time.monotonic()
        print(f"execute={int((t1-t0)*1000)}ms checkpoint={int((t2-t1)*1000)}ms")
    db.close()


if __name__ == "__main__":
    main(n_txns=100, n_rows_per_txn=1_000, n_bytes_per_row=10_000)
```

This prints something like:
```
execute=162ms checkpoint=2853ms
execute=232ms checkpoint=5995ms
execute=381ms checkpoint=9048ms
execute=228ms checkpoint=12221ms
...
```

(Oddly, if I set `n_bytes_per_row=1000` and `n_rows_per_txn=10000`, then it runs significantly faster, even though the total data is the same size. In that scenario, the pattern is that checkpoint time gradually climbs from about 200ms to about 2000ms over the course of about 10 repetitions, then drops back down to 200ms, and the pattern repeats.)

### OS:

macOS Ventura 13.4

### DuckDB Version:

duckdb-0.8.1.dev111

### DuckDB Client:

Python

### Full Name:

Timothy Maxwell

### Affiliation:

Anthropic

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by paultiq** [comment]
fwiw: reproduced, but if you change it to create a new table for each iteration, the checkpoint time is constant. So, it seems more related to the size of the table than the size of the db: 

I changed only the execute call: 
```
db.execute(
            f"""
            create table test{i} as
            SELECT test::varchar
            FROM data
        """
        )
```

**Comment by tmaxwell-anthropic** [comment]
Ah, yes, whoops! I said "database size" but meant "table size". Edited the issue description.

**Comment by paultiq** [comment]
I modified the test to add a little more information: table size and row count, and realized all we're looking at is just that the checkpoint time increases up until STANDARD_ROW_GROUPS_SIZE (122880), which is as expected since the unit of checkpointing is the STANDARD_ROW_GROUPS_SIZE [22-duckdb.pdf](https://15721.courses.cs.cmu.edu/spring2023/slides/22-duckdb.pdf).

(one of the devs/experts can probably confirm this, this is just my non-expert conclusion)

To illustrate with smaller examples: 
#### 10000 rows per iteration @ 10 bytes per row
![image](https://github.com/duckdb/duckdb/assets/104510378/cccfadd9-16c3-4661-8a96-3a89ed42f456)

#### 10000 rows per iteration @ 1000 bytes per row
![image](https://github.com/duckdb/duckdb/assets/104510378/37e54472-92cc-45d0-9b39-f9efc73fd468)

#### Modified Test
```
import random
import duckdb
import pyarrow
import time
from pathlib import Path
import csv 

results = []

def main(*, n_txns, n_rows_per_txn, n_bytes_per_row):
    db_file = Path("./ducktest.duckdb")
    if db_file.exists():
        db_file.unlink()
    with duckdb.connect(str(db_file)) as db:
        db.execute("CREATE TABLE test (test VARCHAR)")
        for i in range(n_txns):
            data = pyarrow.table(
                {
                    "test": pyarrow.array(
                        [random.randbytes(n_bytes_per_row) for i in range(n_rows_per_txn)]
                    )
                }
            )
            t0 = time.monotonic()
            db.execute(
                """
                INSERT INTO test
                SELECT test
                FROM data
            """
            )
            t1 = time.monotonic()
            db.execute("CHECKPOINT")
            t2 = time.monotonic()
            space = db.execute("SELECT format_bytes(COUNT(DISTINCT block_id) * 256000) AS storage_space, count(distinct block_id) FROM pragma_storage_info('test')").df().iloc[0].values
            #print(db.execute("pragma storage_info('test')").df())
            #print(space)
            space = space[0]
            numblocks = space[1]
            numrows = db.execute("SELECT count(*) from test").df().iloc[0].values[0]


            newresult=(f"{n_rows_per_txn} rows x {n_bytes_per_row} bytes", int((t1-t0)*1000), int((t2-t1)*1000),space,numrows, numblocks)
            results.append(newresult)
            print(newresult)


if __name__ == "__main__":
    main(n_txns=100, n_rows_per_txn=10000, n_bytes_per_row=1_000)


    with open("results.csv", 'a', newline='') as csvfile:
        writer = csv.writer(csvfile)

        writer.writerow(("Test", "Execution", "CheckpointTime", "space", "numrows", "numblocks"))
        for row in results:
            writer.writerow(row)
```

**Comment by Mytherin** [comment]
Thanks for the report! As indicated by @paultiq this is somewhat expected as the row group will get re-checkpointed (as changes are made to it). However, the checkpoint time here is excessive. I think this is likely due to the fact that we use `gzip` to compress large strings individually which is very slow. I have a plan to switch this to `zstd` and to combine large strings together to speed this up significantly, which I will hopefully get to in the coming weeks.

**Comment by paultiq** [comment]
Oh, this is interesting. I *thought* that because this data was random, and because `pragma storage_info` reports the column as uncompressed, that compression was not a factor. I had also checked with `pragma force_compression='uncompressed'` but misunderstood the result. 

Given the checkpoint compression of large strings occurs only on VARCHARs, I ran two additional tests using 10k rows x 1 column x 10KB test. 
- Using BLOB instead of VARCHAR gave a 10x speedup: 30s => 3s
- Similarly, using compressible text (via Faker) also gave about the same speedup: 30s => 3s
- Therefore: if OP (@tmaxwell-anthropic) was truly writing incompressing data, perhaps switching to a BLOB... or if not incompressible, perhaps retest using a more realistic data pattern? 

@Mytherin Perhaps large string compression could be disabled for uncompressed columns, regardless of other optimizations?
